File: Data Structures & Algorithms/maximum-product-subarray/submission-7.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        maximum = -100
        dp = [[None, None] for _ in range(len(nums))]
        dp[0][0] = nums[0]
        dp[0][1] = nums[0]
        def opt(i, j):
            nonlocal maximum
            if dp[i][j] is None:
                cand = [opt(i - 1, 0) * nums[i], opt(i - 1, 1) * nums[i], nums[i]]
                dp[i][0] = min(cand)
                dp[i][1] = max(cand)
            maximum = max(maximum, dp[i][1])
            return dp[i][j]
        opt(len(nums) - 1, 0)
        print(dp)
        return maximum

File: Data Structures & Algorithms/house-robber-ii/submission-7.cpp
class Solution {
public:
    int robAux(int houses, int start, vector<vector<int>>& dp, vector<int>& nums) {
        if (houses == 0) {
            dp[houses][start] = 0;
        } else if (dp[houses][start] >= 0) {
            return dp[houses][start];
        } else if (houses == 1) {
            dp[houses][start] = nums[houses - 1 + start];
        } else {
            dp[houses][start] = max(robAux(houses - 1, start, dp, nums), robAux(houses - 2, start, dp, nums) + nums[houses - 1 + start]);
        }
        return dp[houses][start];
    }
    int rob(vector<int>& nums) {
        int houses = nums.size();
        vector<vector<int>> dp(nums.size() + 1, vector<int>(2, -1));
        if (houses == 1) {
            return nums[0];
        }
        return max(robAux(houses - 1, 0, dp, nums), robAux(houses - 1, 1, dp, nums));
    }
};
